File: src/lib/draft.ts
import type { Abbr } from "../../shared/teams.ts";

export interface Draft {
  /** gameId -> picked team (unlocked games only). */
  selections: Record<string, Abbr>;
  /** Rank order of the selected gameIds (most confident first). */
  order: string[];
}

const key = (playerId: string, week: number) => `nflpool.draft.${playerId}.${week}`;

export const emptyDraft = (): Draft => ({ selections: {}, order: [] });

export function loadDraft(playerId: string, week: number): Draft | null {
  try {
    const raw = localStorage.getItem(key(playerId, week));
    if (!raw) return null;
    const d = JSON.parse(raw) as Draft;
    if (!d || typeof d !== "object" || !d.selections || !Array.isArray(d.order)) return null;
    return d;
  } catch {
    return null;
  }
}

/**
 * An empty draft is not a draft — it is the absence of one, so it clears rather than writes.
 *
 * The iOS app had a bug this closes: saving picks clears the store and then resets the draft to
 * empty, and that reset is itself a change, so an empty draft was written straight back over the
 * clear and preferred over the saved picks on the next load. The web seeds from the server rather
 * than from storage so it never showed the symptom, but the same rule belongs on both sides.
 */
export function saveDraft(playerId: string, week: number, draft: Draft): void {
  if (draft.order.length === 0) return clearDraft(playerId, week);
  try {
    localStorage.setItem(key(playerId, week), JSON.stringify(draft));
  } catch {
    /* ignore */
  }
}

export function clearDraft(playerId: string, week: number): void {
  try {
    localStorage.removeItem(key(playerId, week));
  } catch {
    /* ignore */
  }
}

export function toggleSelection(draft: Draft, gameId: string, team: Abbr): Draft {
  const current = draft.selections[gameId];
  const selections = { ...draft.selections };
  let order = draft.order.filter((id) => id !== gameId);
  if (current === team) {
    delete selections[gameId];
  } else {
    selections[gameId] = team;
    order = current ? insertAt(draft.order, gameId, draft.order.indexOf(gameId)) : [...order, gameId];
  }
  return { selections, order };
}

function insertAt(list: string[], id: string, index: number): string[] {
  const without = list.filter((x) => x !== id);
  without.splice(index < 0 ? without.length : index, 0, id);
  return without;
}

export function removeSelection(draft: Draft, gameId: string): Draft {
  const selections = { ...draft.selections };
  delete selections[gameId];
  return { selections, order: draft.order.filter((id) => id !== gameId) };
}

export function moveInOrder(order: string[], from: number, to: number): string[] {
  if (to < 0 || to >= order.length || from === to) return order;
  const next = [...order];
  const [item] = next.splice(from, 1);
  next.splice(to, 0, item!);
  return next;
}

File: ios/TallyKit/Sources/TallyKit/Rules/Draft.swift
import Foundation

/**
 The picks a player is putting together for a week, before they are saved. Mirrors
 `src/lib/draft.ts`: a set of selections (game → team) and the order they will be ranked in,
 most confident first. Kept on the phone per player and week, so a killed app loses nothing.
 */
public struct Draft: Codable, Hashable, Sendable {
    /// gameId → picked team (unlocked games only).
    public var selections: [String: String]
    /// Rank order of the selected gameIds (most confident first).
    public var order: [String]

    public static let empty = Draft(selections: [:], order: [])

    public init(selections: [String: String], order: [String]) {
        self.selections = selections
        self.order = order
    }

    public var isEmpty: Bool { order.isEmpty }

    /// Tapping a team: picks it, swaps to it, or (tapping it again) clears the game.
    public func toggling(gameId: String, team: String) -> Draft {
        var next = self
        let current = selections[gameId]
        if current == team {
            next.selections.removeValue(forKey: gameId)
            next.order.removeAll { $0 == gameId }
        } else {
            next.selections[gameId] = team
            if current == nil {
                next.order.removeAll { $0 == gameId }
                next.order.append(gameId)
            }
        }
        return next
    }

    public func removing(gameId: String) -> Draft {
        var next = self
        next.selections.removeValue(forKey: gameId)
        next.order.removeAll { $0 == gameId }
        return next
    }

    public func moving(from: Int, to: Int) -> Draft {
        guard to >= 0, to < order.count, from >= 0, from < order.count, from != to else { return self }
        var next = self
        let item = next.order.remove(at: from)
        next.order.insert(item, at: to)
        return next
    }
}

/// Per player, per week, on this phone.
public enum DraftStore {
    private static func key(_ playerId: String, _ week: Int) -> String { "tally.draft.\(playerId).\(week)" }

    public static func load(playerId: String, week: Int, defaults: UserDefaults = .standard) -> Draft? {
        guard let data = defaults.data(forKey: key(playerId, week)) else { return nil }
        return try? JSONDecoder().decode(Draft.self, from: data)
    }

    /**
     An empty draft is not a draft — it is the absence of one, so it clears rather than writes.

     This matters more than it looks. Saving picks clears the store and then resets the draft to
     empty, and that reset is itself a change, so it wrote an empty draft straight back over the
     clear. `load` then returned that empty-but-present draft, `seedIfNeeded` took it in preference
     to the picks on the server, and the screen came back with an empty tray — five picks that were
     safely saved, looking lost. Making absence and emptiness the same thing here closes it for
     every caller rather than for the one that happened to trip it.
     */
    public static func save(_ draft: Draft, playerId: String, week: Int, defaults: UserDefaults = .standard) {
        guard !draft.isEmpty else { return clear(playerId: playerId, week: week, defaults: defaults) }
        if let data = try? JSONEncoder().encode(draft) { defaults.set(data, forKey: key(playerId, week)) }
    }

    public static func clear(playerId: String, week: Int, defaults: UserDefaults = .standard) {
        defaults.removeObject(forKey: key(playerId, week))
    }
}
